portfolio/src/main/java/com/google/sps/PuzzleDao.java

```diff
@@ -70,11 +73,20 @@ public Puzzle read(Long puzzleId) {
           name = "";
       }
 
+      String username;
+
+      try {
+          username = entity.getString("name");
+      } catch (DatastoreException e) {
+          username = "";
+      }      
+
       Puzzle puzzle = new Puzzle()
-        .setPuzzleId(entity.getKey().getId())
-        .setImageUrl(entity.getString("imageUrl"))
-        .setDifficulty(difficulty)
-        .setName(name);
+        .create(entity.getKey().getId(), 
```

Comment:
Instead of Puzzle.create() I would add a method to PuzzleDao that converts a Puzzle to an entity and vice versa:
```java
private Puzzle getPuzzleFrom(Entity entity) {
  return new Puzzle()
      .setPuzzleId(entity.getKey().getId())
      .setImageUrl(entity.getString("imageUrl"))
      .setDifficulty(Puzzle.Difficulty.valueOf(entity.getString("difficulty")))
      .setName(entity.contains("name") ? entity.getString("name") : "")
      .setUsername(entity.contains("username") ? entity.getString("username" : "");
}
```
Then you can just do the following in this method:
```java
Entity entity = datastore.get(keyFactory.newKey(puzzleId));
return entity != null ? getPuzzleFrom(entity): null;
```